README.md
# build-name

> Heavily inspired by this repository: https://github.com/fnichol/names

A simple cli tool that generates random hyphenated names in an `adjective-noun` format that can be used to identify builds.

After installation simply run `build-names` in your command line to get a single option.

Example output after running `build-names`:

```
$ build-names
> abject-hands
```

## Installation

`curl -o- https://raw.githubusercontent.com/graingiant/build-names/0.0.2/scripts/install.sh | bash`

or

`wget -qO- https://raw.githubusercontent.com/graingiant/build-names/0.0.2/scripts/install.sh | bash`

After installation, the file binary should be in your `/usr/local/bin` directory. Either restart your terminal application or `source` your profile again.

## Options

```
-i --integer "Append random integer to names generated (Ex: abject-hands-0843)"
-d --date "Append todays date to the names generates (Ex: abject-hands-2022-03-13)"
-n --number "Number of names to generate"
-a --alliterate "Make sure the word pairs are alliterative (Ex: abject-animal)"
```

The options `-i` and `-d` will be read in the order given. For example: `-i -d` would yield `abject-hands-9876-2022-03-13` and `-d -i` would yield `abject-hands-2022-03-13-6877`

## License

Copyright 2022 Alex Caza (graingiant / [alexcaza.com](https://alexcaza.com))

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-03-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-03-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/build-name/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/build-name/compare/0.1.0...0.1.1


test/build_name/core_test.clj
(ns build-name.core-test
  (:require [clojure.test :refer :all]
            [build-name.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/build_name/names/generate.clj
(ns build-name.names.generate
  (:require [clojure.string :as str]
            [clojure.java.io :refer [resource]])
  (:import (java.time LocalDate)))


(defn- get-lines [file]
  (str/split-lines (slurp file)))

(defn- get-random [lst] (get lst (rand-int (count lst))))
(defn- get-random-alliterated [lst starting-letter]
  (let [matched-letter-options (filter #(clojure.string/starts-with? % starting-letter) lst)]
    (first (take 1 (shuffle matched-letter-options)))))

(def nouns (get-lines (resource "data/nouns.txt")))
(def adjectives (get-lines (resource "data/adjectives.txt")))
(def date-formatted (.toString (LocalDate/now)))

(defn- generate-alliterative-name []
  (let [noun (get-random nouns)
        adjective (get-random-alliterated adjectives (subs noun 0 1))]
    (str adjective "-" noun)))

(defn- generate-default-name []
  (let [noun (get-random nouns)
        adjective (get-random adjectives)]
    (str adjective "-" noun)))

(defn generate-name
  "Generates a name based on a random noun + adjective"
  ([_] (generate-alliterative-name))
  ([] (generate-default-name)))

(defn append-id [name] (str name "-" (format "%04d" (rand-int 9999))))

(defn append-date [name] (str name "-" date-formatted))


src/build_name/core.clj
(ns build-name.core
  (:require [clojure.tools.cli :refer [parse-opts]]
            [build-name.names.generate :refer [generate-name append-id append-date]])
  (:gen-class))

(def cli-options [[nil "--version" "Version number" :id :version]
                  ["-i" "--integer" "Append random integer to names generated (Ex: abject-hands-0843)" :id :integer]
                  ["-d" "--date" "Append todays date to the names generates (Ex: abject-hands-2022-03-13)" :id :date]
                  ["-n" "--number NUMBER" "Number of names to generate" :default 1 :parse-fn #(Integer/parseInt %)]
                  ["-a" "--alliterate" "Make sure the word pairs are alliterative" :id :alliterate]])

(def func-map {:date append-date :integer append-id})

(defn apply-modifier [name [option]]
  (apply (get-in func-map [option]) [name]))

(defn generate-names [modifiers]
  (if (contains? modifiers :alliterate)
    (reduce apply-modifier (generate-name true) (apply dissoc modifiers [:alliterate]))
    (reduce apply-modifier (generate-name) (apply dissoc modifiers [:alliterate]))))

(defn -main
  "Generate a random noun-adjective name pair."
  [& args]
  (let [opts (parse-opts args cli-options)
        num-to-gen (get-in opts [:options :number])
        modifiers (apply dissoc (get-in opts [:options]) [:version :number])
        output (repeatedly num-to-gen #(generate-names modifiers))]
    (run! println output)))
